File: frontend/src/api/operatorTaskTruth.ts
/**
 * Canonical operator execution task truth (W6-T01).
 * GET /api/v1/operator/orders/{orderId}/task-truth
 */
import { getAPIBaseURL } from "@/lib/config";

export const OPERATOR_TASK_TRUTH_VERSION = "operator_task_truth/v1";

export type ReadinessAuthority =
  | "FROZEN_ORDER_SNAPSHOT_V2"
  | "LEGACY_ORDER_INPUT"
  | "LEGACY_READ_MODEL_EXPLICIT"
  | "BLOCKED_MISSING_ORDER_SNAPSHOT_V2";

export type IdentitySource =
  | "frozen_task_identity/v1"
  | "legacy_plan_task"
  | "not_proven";

export interface TaskIdentityTruth {
  task_id: string;
  deterministic_task_key?: string | null;
  display_label: string;
  identity_classification?: string | null;
  source_graph_node_id?: string | null;
  source_component_instance_id?: string | null;
  component_role?: string | null;
  component_label?: string | null;
  component_template_code?: string | null;
  source_operation_code?: string | null;
  source_task_rule_code?: string | null;
  parent_graph_node_id?: string | null;
  task_scope?: string | null;
  logo_segment_key?: string | null;
  identity_source: IdentitySource;
}

export interface TaskRuntimeTruth {
  current_status: string;
  assigned_employee_id?: number | null;
  assigned_employee_name?: string | null;
  is_startable: boolean;
  is_completeable: boolean;
  is_blocked: boolean;
  readiness_status?: string | null;
  readiness_label?: string | null;
  readiness_reasons: Array<Record<string, unknown>>;
  blocking_reasons: Array<Record<string, unknown>>;
  blocking_task_ids: string[];
  blocking_tasks: Array<Record<string, unknown>>;
  production_release_blocked: boolean;
  production_release_status: string;
  production_release_scope: string;
  blocking_owner_decision_codes: string[];
  last_started_at?: string | null;
  last_ended_at?: string | null;
}

export interface TaskAuthorityTruth {
  frozen_source?: string | null;
  operational_source: string;
  readiness_source: string;
  production_release_source: string;
  legacy_fallback_active: boolean;
}

export interface OperatorTaskTruthTask {
  identity: TaskIdentityTruth;
  runtime: TaskRuntimeTruth;
  authority: TaskAuthorityTruth;
}

export interface OwnerDecisionSummaryItem {
  code: string;
  label: string;
  category: string;
  blocking: boolean;
  frozen_status: string;
  operational_status: string;
  scope: string;
  required_action?: string | null;
  acknowledgement_sufficient: boolean;
  requires_resolution: boolean;
  can_resolve: boolean;
  resolved_at?: string | null;
  resolved_by_user_name?: string | null;
  has_resolution_note: boolean;
}

export interface RoleCapabilities {
  can_resolve_owner_decisions: boolean;
  can_view_internal_cost: boolean;
  can_view_owner_decision_notes: boolean;
}

export interface InternalCostSummary {
  visibility: "available" | "restricted";
  status?: string | null;
  estimated_total_internal_cost?: number | null;
  accepted_commercial_total?: number | null;
  execution_blocked?: boolean | null;
}

export interface OperatorTaskTruthResponse {
  contract_version: string;
  order_id: number;
  order_code?: string | null;
  execution_plan_id?: number | null;
  order_snapshot_v2_id?: number | null;
  quote_snapshot_v2_id?: number | null;
  task_identity_version?: string | null;
  readiness_authority: ReadinessAuthority;
  production_release_policy: string;
  production_release_status: string;
  production_release_blocked: boolean;
  owner_decisions_summary: OwnerDecisionSummaryItem[];
  role_capabilities: RoleCapabilities;
  internal_cost_summary?: InternalCostSummary | null;
  tasks: OperatorTaskTruthTask[];
  generated_at: string;
  legacy_order: boolean;
}

const operatorBase = () => `${getAPIBaseURL()}/api/v1/operator`;

export async function fetchOperatorTaskTruth(
  orderId: number,
): Promise<OperatorTaskTruthResponse> {
  const res = await fetch(`${operatorBase()}/orders/${orderId}/task-truth`, {
    credentials: "include",
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(
      (err as { detail?: { message?: string } })?.detail?.message ||
        `task-truth ${res.status}`,
    );
  }
  return res.json() as Promise<OperatorTaskTruthResponse>;
}

/** Client must not derive startability — use runtime.is_startable from API only. */
export function taskTruthStartableFromBackend(task: OperatorTaskTruthTask): boolean {
  return task.runtime.is_startable === true;
}

File: backend/schemas/operator_task_truth.py
"""Canonical operator execution task truth read model (W6-T01)."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from schemas.execution_owner_decision_release import PRODUCTION_RELEASE_POLICY
from schemas.execution_plan_v2_frozen_task_identity import FROZEN_TASK_IDENTITY_VERSION

OPERATOR_TASK_TRUTH_VERSION = "operator_task_truth/v1"
PRODUCTION_RELEASE_SCOPE = "ORDER_SCOPE"

ReadinessAuthority = Literal[
    "FROZEN_ORDER_SNAPSHOT_V2",
    "LEGACY_ORDER_INPUT",
    "LEGACY_READ_MODEL_EXPLICIT",
    "BLOCKED_MISSING_ORDER_SNAPSHOT_V2",
]

IdentitySource = Literal[
    "frozen_task_identity/v1",
    "legacy_plan_task",
    "not_proven",
]


class TaskIdentityTruth(BaseModel):
    task_id: str
    deterministic_task_key: str | None = None
    display_label: str
    identity_classification: str | None = None
    source_graph_node_id: str | None = None
    source_component_instance_id: str | None = None
    component_role: str | None = None
    component_label: str | None = None
    component_template_code: str | None = None
    source_operation_code: str | None = None
    source_task_rule_code: str | None = None
    parent_graph_node_id: str | None = None
    task_scope: str | None = None
    logo_segment_key: str | None = None
    identity_source: IdentitySource = "not_proven"


class TaskRuntimeTruth(BaseModel):
    current_status: str
    assigned_employee_id: int | None = None
    assigned_employee_name: str | None = None
    is_startable: bool = False
    is_completeable: bool = False
    is_blocked: bool = False
    readiness_status: str | None = None
    readiness_label: str | None = None
    readiness_reasons: list[dict[str, Any]] = Field(default_factory=list)
    blocking_reasons: list[dict[str, Any]] = Field(default_factory=list)
    blocking_task_ids: list[str] = Field(default_factory=list)
    blocking_tasks: list[dict[str, Any]] = Field(default_factory=list)
    production_release_blocked: bool = False
    production_release_status: str
    production_release_scope: str = PRODUCTION_RELEASE_SCOPE
    blocking_owner_decision_codes: list[str] = Field(default_factory=list)
    last_started_at: str | None = None
    last_ended_at: str | None = None


class TaskAuthorityTruth(BaseModel):
    frozen_source: str | None = None
    operational_source: str = "execution_reality"
    readiness_source: str = "task_readiness_service"
    production_release_source: str = "execution_owner_decision_production_release_service"
    legacy_fallback_active: bool = False


class OperatorTaskTruthTask(BaseModel):
    identity: TaskIdentityTruth
    runtime: TaskRuntimeTruth
    authority: TaskAuthorityTruth


class OwnerDecisionSummaryItem(BaseModel):
    code: str
    label: str
    category: str
    blocking: bool
    frozen_status: str = "present"
    operational_status: str
    scope: str = "order"
    required_action: str | None = None
    acknowledgement_sufficient: bool = False
    requires_resolution: bool = False
    can_resolve: bool = False
    resolved_at: str | None = None
    resolved_by_user_name: str | None = None
    has_resolution_note: bool = False


class RoleCapabilities(BaseModel):
    can_resolve_owner_decisions: bool = False
    can_view_internal_cost: bool = False
    can_view_owner_decision_notes: bool = False


class InternalCostSummary(BaseModel):
    visibility: Literal["available", "restricted"]
    status: str | None = None
    estimated_total_internal_cost: float | None = None
    accepted_commercial_total: float | None = None
    execution_blocked: bool | None = None


class OperatorTaskTruthResponse(BaseModel):
    contract_version: str = OPERATOR_TASK_TRUTH_VERSION
    order_id: int
    order_code: str | None = None
    execution_plan_id: int | None = None
    order_snapshot_v2_id: int | None = None
    quote_snapshot_v2_id: int | None = None
    task_identity_version: str | None = None
    readiness_authority: ReadinessAuthority
    production_release_policy: str = PRODUCTION_RELEASE_POLICY
    production_release_status: str
    production_release_blocked: bool
    owner_decisions_summary: list[OwnerDecisionSummaryItem] = Field(default_factory=list)
    role_capabilities: RoleCapabilities
    internal_cost_summary: InternalCostSummary | None = None
    tasks: list[OperatorTaskTruthTask] = Field(default_factory=list)
    generated_at: str
    legacy_order: bool = False


# Re-export for schema tests
CANONICAL_FROZEN_IDENTITY_FIELDS = (
    "contract_version",
    "deterministic_task_key",
    "source_graph_node_id",
    "source_component_role",
    "source_template_code",
    "source_component_instance_id",
    "source_segment_key",
    "source_operation_code",
    "source_task_rule_code",
    "parent_graph_node_id",
    "operation_scope",
    "identity_classification",
    "task_rule_origin",
)

DEFAULT_TASK_IDENTITY_VERSION = FROZEN_TASK_IDENTITY_VERSION
